service/frontend/adminHandler.go

```diff
@@ -156,47 +160,127 @@ func (adh *AdminHandler) AddSearchAttribute(ctx context.Context, request *admins
 		return nil, adh.error(errRequestNotSet, scope)
 	}
 
-	if len(request.GetSearchAttribute()) == 0 {
+	if len(request.GetSearchAttributes()) == 0 {
 		return nil, adh.error(errSearchAttributesNotSet, scope)
 	}
 	if err := adh.validateConfigForAdvanceVisibility(); err != nil {
 		return nil, adh.error(errAdvancedVisibilityStoreIsNotConfigured, scope)
 	}
 
-	searchAttr := request.GetSearchAttribute()
-	currentValidAttr, _ := adh.DynamicConfigClient.GetMapValue(
-		dynamicconfig.ValidSearchAttributes, nil, searchattribute.GetDefaultTypeMap())
-	for k, v := range searchAttr {
-		if searchattribute.IsReservedField(k) || searchattribute.IsBuiltIn(k) {
-			return nil, adh.error(serviceerror.NewInvalidArgument(fmt.Sprintf(errKeyIsReservedBySystemMessage, k)), scope)
+	indexName := request.GetIndexName()
+	if indexName == "" {
+		indexName = adh.ESConfig.GetVisibilityIndex()
+	}
+
+	currentSearchAttributes, err := adh.Resource.GetSearchAttributesProvider().GetSearchAttributes(indexName, true)
+	if err != nil {
+		return nil, adh.error(serviceerror.NewInternal(fmt.Sprintf(errUnableToGetSearchAttributesMessage, err)), scope)
+	}
+
+	for saName, saType := range request.GetSearchAttributes() {
+		if searchattribute.IsReservedField(saName) {
+			return nil, adh.error(serviceerror.NewInvalidArgument(fmt.Sprintf(errSearchAttributeIsReservedMessage, saName)), scope)
 		}
-		if _, exist := currentValidAttr[k]; exist {
-			return nil, adh.error(serviceerror.NewInvalidArgument(fmt.Sprintf(errKeyIsAlreadyWhitelistedMessage, k)), scope)
+		if currentSearchAttributes.IsValid(saName) {
```

Comment:
`!if currentSearchAttributes.IsValid(saName) {` ?

Reply:
`IsValid` is just a bad name. I am renaming it to `IsDefined` and this `if` checks if search attribute with this name is already defined.